master/buildbot/steps/master.py

```diff
@@ -242,3 +246,134 @@ def start(self):
+        self._pre_process = preProcess
+
+    def get_candidates(self, builder_names):
+        _hush_pyflakes = [builder_names]
+        del _hush_pyflakes
+        raise NotImplementedError
+
+    def handle_candidate(self, control):
+        _hush_pyflakes = [control]
+        del _hush_pyflakes
+        raise NotImplementedError
+
+    @defer.inlineCallbacks
+    def run(self):
+        all_names = self.master.botmaster.builderNames[:]
+
+        if self._builder_names is None:
+            builder_names = all_names
+        else:
+            builder_names = [name for name in self._builder_names in all_names]
+
+        ours = self._pre_process(self.build.sources)
+
+        candidates = yield self.get_candidates(builder_names)
+
+        for control, theirs in candidates:
+            if self._is_relevant(ours, theirs):
+                self.handle_candidate(control)
+
+        # This step can never fail
+        defer.returnValue(SUCCESS)
+
+
+class CancelRelatedBuilds(_HandleRelatedBuilds):
+    name = 'CancelRelatedBuilds'
+    description = ['Checking']
+    descriptionDone = ['Checked']
+
+    def __init__(self, builderNames=None, isRelevant=None, preProcess=None,
+                 **kwargs):
+        _HandleRelatedBuilds.__init__(self, builderNames, isRelevant,
+                                      preProcess)
+
+    @defer.inlineCallbacks
+    def get_candidates(self, builder_names):
+        result = []
+
+        master_control = interfaces.IControl(self.master)
+
+        for name in builder_names:
+            builder_control = master_control.getBuilder(name)
+
+            pending = yield builder_control.getPendingBuildRequestControls()
+
+            # How can it even return None?
+            if pending is None:
+                continue
+
+            for buildrequest in pending:
+                result.append((buildrequest,
+                               [buildrequest.original_request.source]))
+
+        defer.returnValue(result)
+
+    def handle_candidate(self, control):
+        control.cancel()
+
+
+class StopRelatedBuilds(_HandleRelatedBuilds):
+    name = 'StopRelatedBuilds'
+    description = ['Checking']
+    descriptionDone = ['Checked']
+
+    def __init__(self, builderNames=None, isRelevant=None, preProcess=None,
+                 reason=None, **kwargs):
+        _HandleRelatedBuilds.__init__(self, builderNames, isRelevant,
+                                      preProcess)
+
+        if reason is None:
+            reason = 'Stopped by StopRelatedBuilds'
+
+        self._reason = reason
+
+    @defer.inlineCallbacks
+    def get_candidates(self, builder_names):
+        result = []
+
+        master_status = self.master.status
+        master_control = interfaces.IControl(self.master)
+
+        for name in builder_names:
+            builder_status = master_status.getBuilder(name)
+
+            state, current_builds = builder_status.getState()
+
+            if state != 'building':
```

Comment:
I have an old version of buildbot in prod, but I often see builders saying they are idle, while they have current builds, so I'll keep it to
```
if not current_builds:
```